### DISCUSSION TITLE
new "make an empty commit" option is bad UX

### DISCUSSION THREAD

**Discussion by gfim** [gitextensions/gitextensions#11236] [General]
### Environment

- Git Extensions 4.0.1.15887
- Build f2567dea28bbf0e2cbcd2a95d4c7200e47d0d671
- Git 2.39.0.windows.1
- Microsoft Windows NT 10.0.19044.0
- .NET 6.0.12
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions
- Microsoft.WindowsDesktop.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

### Issue description

The new "Confirm commit" dialog added in GitExtensions 4 is not a good user experience - it's too easy to accidentally press the "Make an empty commit" option. This option would be used very rarely and should be harder to select. Maybe a second dialog to "Confirm empty commit". Or make the "Make an empty commit" option as a checkbox.

### Steps to reproduce

- Open a repo
- Make a change to a file in the repo
- Press the Commit button to bring up the Commit window
- Press Commit
- The "Confirm commit" dialog is shown with the options "Stage and commit all unstaged files" and "Make an empty commit"
- Accidentally select "Make an empty commit"

### Did this work in previous version of GitExtensions?

In older versions, the "Make an empty commit" option was not available

**Comment by RussKie** [comment]
Screenshots would be helpful

**Comment by gfim** [comment]
![image](https://user-images.githubusercontent.com/634925/211709145-1f5e9125-d60e-496d-9dc0-afa91fa3ac31.png)

**Comment by vbjay** [comment]
I'm sorry but reading what is being clicked would solve your problem along with a reset branch mixed or soft would undo the commit and leave working directory the same.

**Comment by mstv** [comment]
A local commit can easily be undone as vbjay mentioned.

![image](https://user-images.githubusercontent.com/36601201/211905698-e850a36f-d80f-4d30-861b-9cbabe5e4e1d.png)

![image](https://user-images.githubusercontent.com/36601201/211905842-18273ddd-6cd0-4e56-875c-60a1253cbb5b.png)

In case you want to add files, a commit can be amended.

![image](https://user-images.githubusercontent.com/36601201/211906496-4e63b713-2fb7-4a66-9e80-61e9f75e05bc.png)

**Comment by gfim** [comment]
Yes, it's not a showstopper. And, yes it can be reversed (although not as easily if you've done a "Commit and push" rather than just a "Commit"). But it's still bad UX.

And saying "reading what is being clicked would solve your problem" is a lot of what's wrong with open source software.

**Comment by crypto-rsa** [comment]
To add my two cents, picking an option _which is not selected by default_ in a dialog with a big red error icon on it doesn't really count as "accidental".

**Comment by pvginkel** [comment]
I also think the new user experience is unfortunate. Before this change, the user was asked whether he wanted to stage all his changes. Most of the time, that's what I wanted. Just confirming the dialog did the right thing for me. Now, I have to choose between the very likely option of staging all items, and the option I'm very sure I will never ever want: make an empty commit. I'm noticing that having this dialog in my primary workflow causes slow down. I use Git Extensions over IDE integration because the tool is fast and powerful. This new user experience slows me down regularly.

I would love to get a setting to revert back to the previous user experience where I was asked whether I wanted to stage all files. This new user experience is not a showstopper of course. I just feel like it would be more in line of what I feel Git Extensions' strength lies.

I can create a PR for the change if the team is interested.

**Comment by pmiossec** [comment]
>This new user experience slows me down regularly.

For me, you have exactly the same number of actions to do than before, just a click and you're done.
And as it is the default action, you can even press the space or enter key (which is even quicker than using the mouse).
I don't feel the UX regression you are talking about even if I acknowledge that the "Make and empty commit" is rarely used.

I have the feeling that your discomfort is that you are committing like you did it with TFVC and not the "git way".
You are not using the staging area and are not carefully staging the files or part of files you want to commit to prevent commit of files or piece of code that you don't want to commit or should be part of another commit like I see it too often.

Exceptionally, there is the stage all button but I nearly never use it because I want to be confident about what I commit.

I think it's a better practice to commit like that. It is just a head up and I understand it won't change things for you.

When I do that, I nearly never see this popup and when I see it, it means that I forgot to stage and I'm more interested to click 'cancel' or press "Escape" to close the popup 😉

**Comment by RussKie** [comment]
> I also think the new user experience is unfortunate. Before this change, the user was asked whether he wanted to stage all his changes. Most of the time, that's what I wanted. Just confirming the dialog did the right thing for me. Now, I have to choose between the very likely option of staging all items, and the option I'm very sure I will never ever want: make an empty commit. I'm noticing that having this dialog in my primary workflow causes slow down. I use Git Extensions over IDE integration because the tool is fast and powerful. This new user experience slows me down regularly.

@pvginkel  I'm having troubles understanding the issue. Previously a user was asked to stage all files, if none was staged. Now the dialog asks the same thing as well as provides another option. Where's the slowing down?

**Comment by gfim** [comment]
I don't think that it's really slowing anything down (except maybe my brain!). In my opinion, it's that an option that's used very rarely (empty commit) is just as prominent as the most common option (stage and commit all files). It should be harder to accidentally select the empty commit option - either a confirmation dialog or a checkbox rather than a prominent button. As I said above, I don't think it's a big deal - but the current implementation is a bad user experience.

**Comment by vbjay** [comment]
> I don't think that it's really slowing anything down (except maybe my brain!). In my opinion, it's that an option that's used very rarely (empty commit) is just as prominent as the most common option (stage and commit all files). It should be harder to accidentally select the empty commit option - either a confirmation dialog or a checkbox rather than a prominent button. As I said above, I don't think it's a big deal - but the current implementation is a bad user experience.

It's a taskdialog. If you read, you won't pick wrong option. I 5hink you are complaining over a non issue.

**Comment by pvginkel** [comment]
I think the issue is mostly due to the dialog looking so threatening. You're right, you could read it. But that's basically the issue.

There's another example. If you do "Commit and Push", I know that sometimes I'll get two confirmation dialogs after that. I know what they are ("you're pushing a new branch" and "do you want to set the tracking branch"). The default for those is fine. The only time I need to cancel out of them is when I'm accidentally pushing a new branch, which gives value.

This one is different. You're presented with a a lot of text in a large font and a threatening red icon. So, I read the dialog, every time. Once I'm reading a dialog through, I switch from the mouse to the keyboard. So now instead of just using Enter out of muscle memory like the previous example, I'm in "attention mode" and am using the mouse. Then I misclick, regularly (because I am still in "fast mode", I just want to create a commit 😄). I think that happened 5 to 10 times by now. So I need to commit again, do an amend commit and stage all files. A lot of extra steps.

I understand the arguments from the developers. Maybe it's just a small number of people. I'll adopt the new workflow of course. I like the tool and this is nowhere close to a good enough reason for me to stop using the tool. I'm just trying my best to give an idea of why I feel the new user experience slows me down. I don't see any upvotes on this issue so I must assume we're in the minority.

**Reply by vbjay** [reply]
Frankly you don't get it if you actually stage what you want in the commit. This  commit and stage all files flow is kinda silly. I don't see threatening ui. I see two actions I can perform. I can either create an empty commit or I can add all files and then commit that change. That is perfect for a task dialog and I don't see the problem. 

We have stage all for people that just want to take all changes. So there is no need to get that dialog.

**Comment by gerhardol** [comment]
There were some discussions when adding "empty commit", an additional checkbox would intimidate users more.
Before there was one block of text, now there is two, so I would like to keep this behavior.
